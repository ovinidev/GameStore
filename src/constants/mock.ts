import { Game } from "@interfaces/games";

export const GAME_MOCK: Game = {
	title: "Game Beta",
	description: "A thrilling game",
	genre: "Action",
	id: 1,
	image: "image-url",
	platform: "PC",
	price: 60,
	release_date: "2024-08-01",
};

export const GAMES_MOCK: Game[] = [
	{
		title: "Game Beta",
		description: "A thrilling game",
		genre: "Action",
		id: 1,
		image: "image-url",
		platform: "PC",
		price: 60,
		release_date: "2024-08-01",
	},
	{
		title: "Game Alpha",
		description: "A thrilling game",
		genre: "Action",
		id: 2,
		image: "image-url",
		platform: "PC",
		price: 60,
		release_date: "2024-08-01",
	},
];

export const CART_MOCK: Game[] = [
	{
		title: "Game Beta",
		description: "A thrilling game",
		genre: "Action",
		id: 1,
		image: "image-url",
		platform: "PC",
		price: 60,
		release_date: "2024-08-01",
	},
	{
		title: "Game Alpha",
		description: "A thrilling game",
		genre: "Action",
		id: 2,
		image: "image-url",
		platform: "PC",
		price: 60,
		release_date: "2024-08-01",
	},
];
