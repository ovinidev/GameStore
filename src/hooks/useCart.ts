import { useAddCart, useDeleteCart } from "@mutations/cart";
import { useGetCart } from "@queries/cart";

export const useCart = () => {
	const { data } = useGetCart();
	const { mutateAsync: addToCart } = useAddCart();
	const { mutateAsync: deleteFromCart } = useDeleteCart();

	return {
		addToCart,
		deleteFromCart,
		data,
	};
};
