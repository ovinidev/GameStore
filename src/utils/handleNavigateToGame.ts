import { Router } from "expo-router";

interface HandleNavigateToGameProps {
	router: Router;
	id: number;
}

export const handleNavigateToGame = ({
	router,
	id,
}: HandleNavigateToGameProps) => {
	router.push(`/${id}`);
	router.setParams({
		id,
	});
};
