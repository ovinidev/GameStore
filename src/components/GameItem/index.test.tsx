import { GAME_MOCK } from "@constants/mock";
import { useCart } from "@hooks/useCart";
import { fireEvent, render } from "@testing-library/react-native";
import { handleNavigateToGame } from "@utils/handleNavigateToGame";
import { GameItem } from "./index";

jest.mock("@hooks/useCart", () => ({
	useCart: jest.fn().mockReturnValue({
		addToCart: jest.fn(),
		deleteFromCart: jest.fn(),
	}),
}));

jest.mock("expo-router", () => ({
	useRouter: jest.fn().mockReturnValue({
		push: jest.fn(),
		setParams: jest.fn(),
	}),
}));

jest.mock("@utils/handleNavigateToGame", () => ({
	handleNavigateToGame: jest.fn(),
}));

const useCartMocked = useCart as jest.Mock;

describe("GameItem", () => {
	it("should be able to match the snapshot", () => {
		const { toJSON } = render(
			<GameItem data={GAME_MOCK} isInTheCart={false} />,
		);

		expect(toJSON()).toMatchSnapshot();
	});

	it("should be able to call handleUpdateCart when cart empty and the button is pressed", () => {
		const { getByTestId } = render(
			<GameItem data={GAME_MOCK} isInTheCart={false} />,
		);
		const { addToCart } = useCartMocked();

		const updateCartButton = getByTestId("update-cart-button");
		fireEvent.press(updateCartButton);

		expect(addToCart).toHaveBeenCalledWith(GAME_MOCK);
	});

	it("should be able to call handleUpdateCart when the item has in cart and the button is pressed", () => {
		const { getByTestId } = render(<GameItem data={GAME_MOCK} isInTheCart />);
		const { deleteFromCart } = useCartMocked();

		const updateCartButton = getByTestId("update-cart-button");
		fireEvent.press(updateCartButton);

		expect(deleteFromCart).toHaveBeenCalledWith(GAME_MOCK.id);
	});

	it("should be able to call handleNavigateToGame with correct parameters when pressed", () => {
		const { getByTestId } = render(<GameItem data={GAME_MOCK} isInTheCart />);
		const touchableNavigate = getByTestId("game-item-touchable");

		fireEvent.press(touchableNavigate);

		expect(handleNavigateToGame).toHaveBeenCalledWith({
			id: GAME_MOCK.id,
			router: expect.any(Object),
		});
	});
});
