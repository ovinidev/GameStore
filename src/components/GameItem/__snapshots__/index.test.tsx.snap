// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameItem should be able to match the snapshot 1`] = `
<View
  accessibilityLabel="Adicionar ao carrinho"
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
    }
  }
>
  <Image
    className="h-[170] w-full"
    resizeMode="cover"
    source={
      {
        "uri": "image-url",
      }
    }
  />
  <Text
    className="py-2 text-center font-semibold text-md text-slate-50"
  >
    Game Beta
  </Text>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
  >
    <Text
      className="text-center font-bold text-md text-slate-50"
    >
      Adicionar
    </Text>
  </View>
</View>
`;
